modules/querying/pages/query-optimizer/enable-cost-optimizer.adoc

```diff
@@ -0,0 +1,108 @@
+
+* Cardinality
+** Vertex types
+*** `Tag`
+*** `Comment`
+*** `Person`
+** Refined edge types
+*** `HAS_TAG`: from `Comment` to `Tag`
+*** `HAS_TAG_REVERSE`: from `Tag` to `Comment`
+*** `LIKES`: from `Person` to `Comment`
+*** `LIKES_REVERSE`: from `Comment` to `Person`
+* Histogram
+** `gender` attribute of type `Person`
+** `id` attribute of type `Tag`
+
+
+=== Compute cardinality data
+To compute cardinality data (the count of each type), use the xref:query-optimizer/stats-api.adoc#_compute_cardinality_statistics[`POST :14240/gsqlserver/gsql/stats/card` endpoint].
+
+For example, to compute the cardinality data of the vertex `Person` in graph `ldbc_snb`, make the following request:
+
+[source.wrap,console]
+----
+curl -s --user tigergraph:tigergraph -X POST "localhost:14240/gsqlserver/gsql/stats/card?graph=ldbc_snb&vertex=Person"
+----
+
+For another example, to compute the cardinality data of the refined edge type `LIKES` from a `Person` vertex to a `Comment` vertex, make the following request:
+
+[.wrap,console]
+----
+curl -s --user tigergraph:tigergraph -X POST "localhost:14240/gsqlserver/gsql/stats/card?graph=ldbc_snb&edge=LIKES&from=Person&to=Comment"
+----
+
+For best results, compute the cardinality data for every type you identified in the previous step, each with its own request.
+
+=== Compute histogram data
+To compute histogram data, use the xref:query-optimizer/stats-api.adoc#_compute_histogram_statistics[`POST :14240/gsqlserver/gsql/stats/histogram` endpoint].
+
+For example, to compute the histogram data for attribute `gender` of vertex type `Person` on graph `ldbc_snb`, make the following request:
+
+[.wrap,console]
+----
+$ curl -s --user tigergraph:tigergraph -X POST "http://localhost:14240/gsqlserver/gsql/stats/histogram?graph=ldbc_snb&vertex=Person&attribute=gender&buckets=256"
+----
+
+For best results, compute the histogram data for every vertex attribute referenced in a `WHERE` clause of a `SELECT` statement.
+
+[#_install_query_with_cost_option]
+=== Install query with `-cost` option
+
+Now that the relevant statistics are available to the query optimizer, you can install the query using the cost-based query optimizer.
+To use the cost-based optimizer during installation, use the xref:query-operations.adoc#_install_query[`-cost` option] when installing the query.
+
+For example, to install the query `example3`, run `INSTALL QUERY example3 -cost`.
+
+Alternatively, set xref:ddl-and-loading:system-and-language-basics.adoc#_session_parameters[the `cost_opt` session parameter] to true by running `set cost_opt = true` in the GSQL shell.
+This enables the query optimizer for the rest of your GSQL session.
+You no long need to supply the `-cost` option when installing queries to use the optimizer if the `cost_opt` parameter is set to true.
+
+If you missed any cardinality or histogram data, the optimizer warns you about which types are missing, so you can compute the data for any missing type or attribute.
```

Comment:
It's quite normal to repeat information wherever it's relevant. Users never read the docs from beginning to end - we just want them to see the information wherever it could possibly be relevant.